### PR TITLE
Conditional vs Interventional Distribution

### LINKED ISSUE #551: Conditional vs interventional distributions
# Notebook proposal
**Title**: Conditional vs interventional distributions

## Why should this notebook be added to pymc-examples?
We only have a small number of notebooks covering causal inference topics so far. Causal inference is gaining more and more attention, so showcasing that PyMC can deal with Bayesian Causal inference is important. This particular notebook compares the conditional and interventional distributions and showcases the amazing new `do` operator. This is currently in `pymc-experimental` but will soon be moved into the main `pymc` repo.

This notebook will _not_ compare and contrast interventions and counterfactuals. This will be left for another notebook.

Suggested categories:
* Level: beginner/intermediate
* [Diataxis](https://diataxis.fr/) type: explanation

## Related notebooks
None. But it will join the few existing notebooks in the Causal Inference subsection

# References
* [Causal Inference 2: Illustrating Interventions via a Toy Example](https://www.inference.vc/causal-inference-2-illustrating-interventions-in-a-toy-example/) by Ferenc Huszár.
* Almost certainly more, such as books by Pearl.

### PR DESCRIPTION
**Pull request by drbenvincent**
This PR will add a new example to the causal section of the docs. It focuses on the do operator, and contrasts conditional distributions to interventional distributions. Part of it is heavily inspired by the blog post [Causal Inference 2: Illustrating Interventions via a Toy Example](https://www.inference.vc/causal-inference-2-illustrating-interventions-in-a-toy-example/) by Ferenc Huszár.

* This PR will close #551 
* Conforms to style guide

FYI

* This PR also spawned this issue: https://github.com/arviz-devs/arviz/issues/2258

### CODE REVIEW

File: examples/causal_inference/interventional_distribution.ipynb
Comment (juanitorduz) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>Maybe "The joint distribution could be described as: " -&gt; "The joint distribution can be factorized (or expressed) as: "</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (juanitorduz) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>Shall we generate the random samples via pymc XD (https://www.pymc-labs.io/blog-posts/simulating-data-with-pymc/()</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (juanitorduz) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.        idata1 = pm.sample_prior_predictive(samples=N)
</pre><p>it would be nice to pass the <code>rng</code> to all the sample stems in the notebook to make it fully reproducible. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

**Review by juanitorduz — APPROVED** [review]
Very cool! I left some minor comments!

File: examples/causal_inference/interventional_distribution.ipynb
Comment (larryshamalama) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>Would it be more clear to have $P(x_3 = 1)$ on the RHS and say in text that $P(x_3 = 1) = 1$ in the interventional graph?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (larryshamalama) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>What are the true data-generating parameters that are being estimated here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

Comment (larryshamalama) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>Typo: acutally -&gt; actually</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.ipynb:1]: <p>Thanks for keeping me not lazy. I've added more detail about factorisation and truncated factorisation</p><p> </p>

Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.ipynb:1]: <p>This code is not actually executed, they are just code blocks within markdown. I did think of doing that, but the PyMC way of doing it is perhaps less clear to people. I am totally willing to add the PyMC way of doing it if we think it's important to present a coherent front, but it is a little more verbose.</p>

Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.ipynb:1]: <p>good idea, I missed this. Fixed</p>

Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.ipynb:1]: <p>I think I see where you are coming from. What I've got matched up with section 1.3 of Pearl's Causality book. I think it makes more sense this way because the truncated factorisation eliminates terms.</p>

Comment (NathanielF) [reply examples/causal_inference/interventional_distribution.ipynb:1]: <p>Yeah, i like the emphasis on factorized or decomposed given the uses of the joint distributions we're going to see below. </p>

Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.ipynb:1]: <p>In a previous edit, I've added more mathematical detail about the factorisation, so hopefully this should be clear to people.</p>

Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.ipynb:1]: <p>fixed :)</p>

**Comment by drbenvincent** [comment]
Thanks for the comments @juanitorduz, @larryshamalama, @NathanielF. I think I'm up to date with edits so far. Happy if there's any more feedback, but this is already improved :)

File: examples/causal_inference/interventional_distribution.ipynb
Comment (NathanielF) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>I wonder if this introduction kind of buries the lede a bit. It's the introduction of the famous 'do'-operator. Makes PyMC a direct player in the Causal inference space. Feels like the intro should be punchier. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (NathanielF) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>Similarly a bit quick into the mechanics of the do-operator as a graph mutilation device. Better to set up the idea more slowly. Lead the audience to the idea from : it'd be great to examine counterfactuals to counterfactuals can be seen as interventions on a network graph of causal flow. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (NathanielF) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>That's a great blog post! </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (NathanielF) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>I'd maybe sign-post a bit more about why this will be important. You don't want to assume the reader has already read the blog-post or that they need to go read the blog post to understand what's happening next. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (NathanielF) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>I think you're moving to a new section here, worth calling out the movement. </p><p> </p><p>"With all this machinery established, we can now clearly illustrate the difference between conditional and interventional distributions using the do-operator" </p><p> </p><p>Kind of a link to the opening paragraph and a signal to the reader that we're getting to the meat of the post. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (NathanielF) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>I'd add a bit of dramatic flair and advertise or hint at potential use-cases. We can see now why causal reasoning is both powerful and important for the implementation of policy and as an aid to counterfactual reasoning. Link back to the opening paragraph. Close with a strong recommendation that this distinction (intervention v conditional distributions needs to be part of the tool-kit of any working data scientist. </p><p> </p><p>Overall the post is great! This is such an important distinction and having the tools to make it well is huge! My comments above are minor stylistic points at most. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.ipynb:1]: <p>Just for context, we have a much punchier announcement of the do operator that will appear as a PyMC Labs blogpost. So this example is less of an announcement than a resource which will hopefully be useful for some time to come. That said, I agree that it's currently a bit meek. I'll up the level of excitement.</p><p> </p>

Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.ipynb:1]: <p>Done. I've added quite a bit of extra text to help with this</p>

Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.ipynb:1]: <p>Done</p>

Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.ipynb:1]: <p>Done</p>

Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.ipynb:1]: <p>Got it - good advice. Maybe not perfect, but I've given this a go</p>

**Comment by drbenvincent** [comment]
Thanks for the feedback everyone. I believe I've dealt with all the feedback to this point. In dealing with that I have added a fair amount of additional introductory and explanatory text. Hopefully this is much improved, but open to any last bits of feedback, particularly on the new content.

File: examples/causal_inference/interventional_distribution.ipynb
Comment (NathanielF) [reply examples/causal_inference/interventional_distribution.ipynb:1]: Looks great now!

Comment (larryshamalama) [reply examples/causal_inference/interventional_distribution.ipynb:1]: I like the new section!! However, I must admit that I tend to prefer more formulas then less, which may or may not be the best for introductory material...

**Review by OriolAbril — COMMENTED** [review]
I think all the converting to dataframe is much more complicated than it needs to be, I'll try running locally and opening a PR against your fork with the proposed changes.

File: examples/causal_inference/interventional_distribution.myst.md
Comment (OriolAbril) [inline examples/causal_inference/interventional_distribution.myst.md:46]: we should also add daft here following https://www.pymc.io/projects/docs/en/latest/contributing/jupyter_style.html#extra-dependencies, not sure about pymc-experimental

File: examples/causal_inference/interventional_distribution.myst.md
Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.myst.md:191]: will fix

File: examples/causal_inference/interventional_distribution.myst.md
Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.myst.md:77]: Will remove `sns.color_palette("tab10")`

File: examples/causal_inference/interventional_distribution.myst.md
Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.myst.md:142]: I'm a bit confused, I can only see catastrophically

File: examples/causal_inference/interventional_distribution.myst.md
Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.myst.md:134]: will try this for all references to notebooks

**Comment by ricardoV94** [comment]
Would it make sense to do some inference? Or at least mention things would be the same if you had done inference, but you would use `do` + `posterior_predictive`

File: examples/causal_inference/interventional_distribution.ipynb
Comment (twiecki) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>DAGs</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

**Comment by drbenvincent** [comment]
> Would it make sense to do some inference? Or at least mention things would be the same if you had done inference, but you would use `do` + `posterior_predictive`

I did think of doing this. But I felt that it would be best to leave that for a future notebook, mainly because this might be about the right volume of content for a beginner to fully digest. I have one planned (will put an issue in soon) which specifically compares interventions vs counterfactuals as Pearl makes a big distinction about these.

File: examples/causal_inference/interventional_distribution.myst.md
Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.myst.md:46]: I've given that a go

File: examples/causal_inference/interventional_distribution.ipynb
Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.ipynb:1]: <p>fixed all instances: DAGS -&gt; DAGs</p>

Comment (twiecki) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>DAGs</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (twiecki) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>DAGs</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (twiecki) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>DAGs</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='43'/>

Comment (twiecki) [inline examples/causal_inference/interventional_distribution.ipynb:1]: <p>mutated? mutilated?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/552/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

Comment (drbenvincent) [reply examples/causal_inference/interventional_distribution.ipynb:1]: Very good point. I think mutation is probably the one to go for. Even though I like the idea of mutilating my DAGS

**Comment by drbenvincent** [comment]
I think we're good to go! Thanks for the feedback everyone - truly valuable. I just need an approval from @twiecki or @OriolAbril before I can merge.

**Review by OriolAbril — APPROVED** [review]

**Pull request merged**
